{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download the E.coli genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "if not os.path.exists(f'{os.getcwd()}/e_coli_genome.fa'):\n",
    "\tsubprocess.run(f'wget -qO- https://ftp.ensemblgenomes.ebi.ac.uk/pub/bacteria/release-62/fasta/bacteria_26_collection/escherichia_coli_w_gca_000184185/dna/Escherichia_coli_w_gca_000184185.ASM18418v1_.dna.toplevel.fa.gz | gunzip > e_coli_genome.fa', shell = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define a function that takes as input a fasta file and a k value and treturns a list of nonmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def nonmers(genomefile, k):\n",
    "    #Example run\n",
    "    # list_of_nonmers = nonmers(path_to_fasta_file, k = 5)\n",
    "    \n",
    "    import regex as re\n",
    "    import itertools\n",
    "\n",
    "    # Importing sequence\n",
    "    file = open(genomefile, 'r')\n",
    "    seq = \"\" #An empty string\n",
    "    count = 0\n",
    "    k =int(k)\n",
    "    for line in file:\n",
    "        count +=1\n",
    "        if (count > 1) : #the first line of a fasta does not contain any part of the sequence\n",
    "            length=len(line)\n",
    "            seq=seq+line[0:length-1] #length-1 because the last character is the \\n character\n",
    "            \n",
    "    file.close()\n",
    "\n",
    "    seq = re.sub(\"[^AGCT]\", \"\", seq) #Remove any non AGCT nucleotide from the sequence\n",
    "\n",
    "    # Creating the kmer combination table\n",
    "    nucleotides = ['A', 'C', 'G', 'T']\n",
    "    allkmers = [''.join(p) for p in itertools.product(nucleotides, repeat=k)] #Create a list of all possible kmers\n",
    "\n",
    "    kmer_genome_list = []\n",
    "    for i in range(len(seq) - k + 1): #slide the whole genome with a step of 1 and append each kmer to the kmer_genome_list\n",
    "        kmer_genome_list.append(seq[i:i+k]) \n",
    "\n",
    "    kmer_genome_set = set(kmer_genome_list) #Keep only the uniq kmers.\n",
    "    return list(set(allkmers) - kmer_genome_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "e_coli_nonmers = nonmers(f'{os.getcwd()}/e_coli_genome.fa', 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'{os.getcwd()}/e_coli_genome.fa', 'r') as genome:\n",
    "    fasta = ''.join(genome.readlines()).strip('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importing sequence\n",
    "file = open(f'{os.getcwd()}/e_coli_genome.fa', 'r')\n",
    "seq = \"\" #An empty string\n",
    "count = 0\n",
    "k =int(10)\n",
    "for line in file:\n",
    "\tcount +=1\n",
    "\tif (count > 1) : #the first line of a fasta does not contain any part of the sequence\n",
    "\t\tlength=len(line)\n",
    "\t\tseq=seq+line[0:length-1] #length-1 because the last character is the \\n character\n",
    "\t\t\n",
    "file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "result = []\n",
    "for nonmer in e_coli_nonmers:\n",
    "    result.append(nonmer in seq)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
