{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "F5-AUY3KZ57N"
      },
      "source": [
        "### PWM -> Position weight matrix. A matrix where eaach entry corresponds to one base of a motif assigned with the frequency found in the motif.\n",
        "\n",
        "### For example if a transcription factor recognizes a motif of 8 nucleotides the a PWM could be the following\n",
        "![image.png](attachment:image.png)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "YzWy6oIvZ57Q"
      },
      "source": [
        "### PSSM -> Position Speciic Scoring Matrix\n",
        "### Is a PWM that takes into acount the background frequency of the nucleotides\n",
        "Each entry on the PSSM is the log2 ratio of the frequency of the nucleotide in the motif and the background frequency of the nucleotide\n",
        "![image.png](attachment:image.png)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4pcRjCefZ57R"
      },
      "source": [
        "## Shannon Entropy\n",
        "![image.png](attachment:image.png)\n",
        "\n",
        " Entropy maximizes when the all possibilities are equally likely to occure.\n",
        " Shannon Information is the difference of the initial entropy and the final entropy\n",
        " This information is measured in bits"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-6fx0OaQZ57S"
      },
      "source": [
        "The most dificult is when I do not have any information about the motif and I am trying to discover it\n",
        "For denovo motif discovery there are approaches with randomized algorithms(Gibbs sampling) and comparative genomicd approaches."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "QnzEcM8mZ57T"
      },
      "source": [
        "Comparative genomics approach -> cross-species sequence alignment to find phylogenetic footprints, regions of unusually high conservation."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "-1vP72u2Z57T"
      },
      "outputs": [],
      "source": [
        "gata = ['TTATAG', 'AGATAT', 'AGATAG', 'ATATCT', 'AGATAG', 'AGATAG', 'AGATAG', 'TGATAA', 'AGATAA',\n",
        " 'AGATAA', 'CGATAG', 'AGAGTT', 'TGATAA', 'TGATAA', 'AGATGG', 'AGATAG', 'AGATTG', 'AGATAA', 'TGATAA',\n",
        "  'AGATAA', 'AGATAG', 'TGATAG', 'TGATCA', 'TTATCA', 'AGATGG', 'TGATAT', 'AGATAG', 'TGATAA', 'GGATAC',\n",
        "   'AGATAA', 'CGATAA', 'TGATAG', 'AGATAA', 'TGATTA', 'AGATAA', 'AGATAG', 'TGATAT', 'AGATAA', 'TCAGAG',\n",
        "    'AAGTAG', 'AGATTA', 'TGATAG', 'TGATAG', 'AGATAC', 'TGATTG', 'AGATTA', 'AGAATA', 'AGATAA', 'AGATTA',\n",
        "     'AGCTTC']\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "wu4Faa2EZ57U",
        "outputId": "23e3e742-a296-4178-ac1e-21c57f9bf78e"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "50"
            ]
          },
          "execution_count": 18,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "len(gata)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XjJz4kkxZ57V"
      },
      "source": [
        "okur"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4yAxlcn3Z57V"
      },
      "source": [
        "A function to create a pwm from a list of aligned sequences"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "2wIv8UXQZ57V"
      },
      "outputs": [],
      "source": [
        "def pwm(sequnce_list):\n",
        "    nuc = ['A', 'T', 'G', 'C']\n",
        "    profile=[[0 for i in range(len(sequnce_list[0]))] for j in range(len(nuc))] #Two list comprehensions\n",
        "    #The first one returns a list of zeros with length equal to the length of each sequence.\n",
        "    #The second one returns 4 lists of length equal to the length of the sequence\n",
        "    #return profile\n",
        "    #\n",
        "    for instance in sequnce_list: #iterates through the sequences\n",
        "        for j in range(len(instance)): #iterates through the indexes of the nucleotides of each sequence\n",
        "            residue = instance[j] #sets the residue variable\n",
        "            profile[nuc.index(residue)][j] +=1 #Adds one to the residue value in the nested list.\n",
        "            profile[nuc.index(residue)][j]=float(profile[nuc.index(residue)][j]) #Converts the previously added value to float\n",
        "    #return profile\n",
        "    import numpy as np\n",
        "    pwm = np.array(profile)\n",
        "    pwm = pwm/len(sequnce_list)\n",
        "    return pwm"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "AafLe4HhZ57W",
        "outputId": "db849d7f-932e-4ed5-9750-4becbb83d153"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "[[0.6  0.02 0.96 0.02 0.72 0.44]\n",
            " [0.34 0.06 0.   0.94 0.18 0.1 ]\n",
            " [0.02 0.9  0.02 0.04 0.04 0.4 ]\n",
            " [0.04 0.02 0.02 0.   0.06 0.06]]\n"
          ]
        }
      ],
      "source": [
        "print(pwm(sequnce_list=gata))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "tLPTJKKuZ57W"
      },
      "outputs": [],
      "source": [
        "#Write a function that reads a fasta sequence\n",
        "def readfasta(fastafile):\n",
        "    import regex as re\n",
        "    f=open(fastafile,'r')\n",
        "    seq = ''\n",
        "    total = 0\n",
        "    for line in f:\n",
        "        x = re.match('>', line)\n",
        "        if x == None:\n",
        "            length = len(line)\n",
        "            total+=length\n",
        "            seq+=line[0:length-1]\n",
        "    return seq\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "qHqm6ZISZ57X"
      },
      "outputs": [],
      "source": [
        "#Write a function that calculates nucleotiude frewuences\n",
        "def nuccomp(sequence):\n",
        "    import numpy as np\n",
        "    nucfreq = [0,0,0,0]\n",
        "    nuc = ['A', 'T', 'G', 'C']\n",
        "    for i in range(len(nuc)):\n",
        "        nucfreq[i] = sequence.count(nuc[i])\n",
        "    nucfreq = np.array(nucfreq)/len(sequence)\n",
        "    return nucfreq"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "lpxALYoDZ57X"
      },
      "outputs": [],
      "source": [
        "seq = readfasta('./Staaur.fa')\n",
        "my_pwm = pwm(sequnce_list=gata)\n",
        "nuc_freq = nuccomp(seq)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "L_n50hCEZ57X",
        "outputId": "7df34d1a-df24-4ce2-f985-6ef651cc1c93"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "array([0.33585935, 0.33495328, 0.16384496, 0.16534241])"
            ]
          },
          "execution_count": 24,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "nuc_freq"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zKkNRnksZ57X"
      },
      "source": [
        "Now that I have computed both nucleotide frwuences and the PWM I can combine them to build a PSSM\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "xWF3sCcNZ57Y"
      },
      "outputs": [],
      "source": [
        "def pssm(pwm, nucfreqs):\n",
        "    import numpy as np\n",
        "    import math\n",
        "    pseudocount = 0.01 #This is very important because if iwant to compute the frequency of any sequence i want to avoid having zeros\n",
        "    pssm = [[0 for i in range(len(pwm[0]))] for j in range(len(nucfreqs))] #This line creates an empty 4 by len(sequence) pssm\n",
        "    for i in range(len(nucfreqs)): #The range is from 1 to 4\n",
        "        pssm[i]=(pwm[i]+pseudocount)/nucfreqs[i] #pssm[i] is the ith row of the pssm array\n",
        "        #The above line takes each row of the pwm matrix and devides it with the nuc_freq array\n",
        "    for i in range(len(pssm)):\n",
        "        for k in range(len(pssm[0])):\n",
        "            pssm[i][k] = math.log(pssm[i][k]/math.log(2))\n",
        "\n",
        "\n",
        "    return np.array(pssm)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "xL69RzUbZ57Y"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "D3KKH-GDZ57Y"
      },
      "outputs": [],
      "source": [
        "my_pssm = pssm(my_pwm, nuc_freq)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "KwF8YQIoZ57Y",
        "outputId": "2509e3a7-f102-410e-c6b5-bb0b97d09153"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "array([[ 0.96327942, -2.04898216,  1.42711653, -2.04898216,  1.14286499,\n",
              "         0.65906804],\n",
              "       [ 0.41045501, -1.1989829 , -3.14489305,  1.40898384, -0.20045407,\n",
              "        -0.74699777],\n",
              "       [-1.33121032,  2.0810369 , -1.33121032, -0.8203847 , -0.8203847 ,\n",
              "         1.28374946],\n",
              "       [-0.82948262, -1.34030824, -1.34030824, -2.43892053, -0.49301038,\n",
              "        -0.49301038]])"
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "my_pssm"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "5mBGtqs-Z57Z",
        "outputId": "f4213b86-e946-47f7-b83c-6e8422032be2"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "array([ 0.09436466, -3.47188893, -0.93840367, -6.9350404 ])"
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "import numpy as np\n",
        "np.sum(my_pssm, axis = 1)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "niEeCpYKZ57Z",
        "outputId": "7315bd00-3a9b-41bc-8d59-382380d51797"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "0.0"
            ]
          },
          "execution_count": 15,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "import math\n",
        "math.log(1,2)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "3c8z6CbqZ57Z",
        "outputId": "6fa3bf8f-0c4a-481f-b505-2329a1bb0678"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "array([-2.04898216, -1.1989829 ,  2.0810369 , -1.34030824])"
            ]
          },
          "execution_count": 28,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "my_pssm[:,1]"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "kTTtaTOkZ57Z"
      },
      "source": [
        "Now that I have also computed a PSSM I can write a function to search in any sequence using this specific PSSM to look for motifs. The probability of any sequence is the producto of the frequnces at each position of the sequence. I have to set a score threshold for that.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Wdzm5LO-Z57Z"
      },
      "outputs": [],
      "source": [
        "def pssmSearch(pssm, sequence, threshold):\n",
        "    nuc = ['A', 'T', 'G', 'C']\n",
        "    hits = []\n",
        "    instances = []\n",
        "    x = [] #A list that holds indexes for the second step\n",
        "    allscores = [] #for plotting reasons\n",
        "\n",
        "    # Step 1: Calculation of the maximum possible pssm score\n",
        "    maxPssm = 0\n",
        "    for j in range(len(pssm[0])): #This essentially gives me the lenth of each sequence.\n",
        "        maxPssm = maxPssm + max(pssm[:,j])\n",
        "        #This means add to max the maximum value of the jth column pssm[:,j] is the jth column of the pssm array\n",
        "        #Note that although I am computing probability I am using the sum of the pssm values and not their product.\n",
        "        #This is because at the last step of the pssm function I log transformed the frequences\n",
        "\n",
        "    # Step 2: Search\n",
        "    for i in range(len(sequence) - len(pssm[0])):\n",
        "        x.append(i)\n",
        "        instance = sequence[i:i+len(pssm[0])]\n",
        "        score = 0\n",
        "        for l in range(len(instance)):\n",
        "            score = score + pssm[nuc.index(instance[l])][l]\n",
        "            if (score > threshold*maxPssm):\n",
        "                hits.append(i)\n",
        "                instances.append(instance)\n",
        "                allscores.append(score)\n",
        "    return (hits, instances, allscores)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "lNi9vmjVZ57a",
        "outputId": "a61ba775-c148-42ee-ccfd-dab1a159c2b1"
      },
      "outputs": [
        {
          "ename": "NameError",
          "evalue": "name 'my_pssm' is not defined",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "Cell \u001b[0;32mIn[14], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m pssmSearch(my_pssm, seq, \u001b[38;5;241m0.80\u001b[39m)[\u001b[38;5;241m1\u001b[39m]\n",
            "\u001b[0;31mNameError\u001b[0m: name 'my_pssm' is not defined"
          ]
        }
      ],
      "source": [
        "pssmSearch(my_pssm, seq, 0.80)[1]\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "M3l6QH8FZ57a"
      },
      "outputs": [],
      "source": [
        "def pwmEntropyInformation(pwm):\n",
        "    import numpy as np\n",
        "    k = pwm.shape[1] #This gives me the number of columns(So the length of the sequence represented in the pwm)\n",
        "    information = np.zeros([1,k]) #Computing the information in each  position\n",
        "    for i in range(k):\n",
        "        information[0,i] = 2 - abs(sum([elem*np.log2(elem) for elem in pwm[:,i] if elem > 0])) #This formula is explained on the top of the document.\n",
        "        sumInfo = np.sum(information)\n",
        "        scaledSumInfo = sumInfo/k\n",
        "\n",
        "    return (information, sumInfo, scaledSumInfo)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "5oDsIpQ6Z57a",
        "outputId": "553526df-72f0-4106-ccf2-d308567724fa"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "0.7300155336155016\n"
          ]
        }
      ],
      "source": [
        "import numpy as np\n",
        "H_init = 2\n",
        "H_final = 0\n",
        "for i in my_pwm[:,0]:\n",
        "    H_final += i*np.log2(i)\n",
        "\n",
        "print(H_init - abs(H_final))\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "HK5gfsbiZ57a",
        "outputId": "3c67e16d-284f-4190-a960-feeea99c5606"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "array([[0.6 , 0.02, 0.96, 0.02, 0.72, 0.44],\n",
              "       [0.34, 0.06, 0.  , 0.94, 0.18, 0.1 ],\n",
              "       [0.02, 0.9 , 0.02, 0.04, 0.04, 0.4 ],\n",
              "       [0.04, 0.02, 0.02, 0.  , 0.06, 0.06]])"
            ]
          },
          "execution_count": 47,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "my_pwm\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "oTq4srIMZ57a",
        "outputId": "c54a5760-ac98-4f2f-b56b-3c6e9118726a"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "(array([[0.73001553, 1.39390935, 1.71770781, 1.61745733, 0.78417406,\n",
              "         0.37435552]]),\n",
              " 6.617619603779856,\n",
              " 1.102936600629976)"
            ]
          },
          "execution_count": 60,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "pwmEntropyInformation(my_pwm)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "QLdKKIdoZ57b",
        "outputId": "5088f579-f39f-480f-c8d5-88d57f8d6943"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "3336"
            ]
          },
          "execution_count": 45,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "len(pwmsearches)\n",
        "len(pssmsearches)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "tdzcwMfsZ57b",
        "outputId": "6616f663-a9b0-4a60-acd4-0a682ffe5486"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "array([[0., 0.]])"
            ]
          },
          "execution_count": 31,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "import numpy as np\n",
        "np.zeros([1,2])"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ATmpsV34Z57b"
      },
      "source": [
        "# Gibbs Sampling"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "C1eM4QwaZ57b"
      },
      "source": [
        "Problem -> Given p strings and a length k, find the most mutually similar length-k substring from each string\n",
        "Gibbs sampling replaces a single k-mer at each iteration and thus moves with more caution in the space of all motifs.\n",
        "“Gibbs sampling” is the basis behind a general class of algorithms\n",
        "that is a type of local search.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cjspsi3tZ57c"
      },
      "source": [
        "Gibbs Sampling do not gurantees good performance but often works well in practice"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "kH4O-29PZ57c"
      },
      "source": [
        "Assumes that we know the length of the motif we are looking for\n",
        "Assumes that each input sequence contains exactly 1 real instance of the motif"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "sfyWJvkhZ57c"
      },
      "source": [
        "Gibbs Sampling Algorithm\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "na-JSpOiZ57c"
      },
      "source": [
        "• Initialization: select random locations in sequences x1, …, xN, and compute an initial\n",
        "model M from these locations"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_Azfof5zZ57d"
      },
      "source": [
        "I divid all sequences to kmers\n",
        "I take one kmer from each sequence and create a pwm\n",
        "I scan all the sequences with the pwm and I substitute the chosen kmer with the one that best fits the pwm\n",
        "Update the PWM\n",
        "Repeat until reaching a set threshold"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Dz7T-KdeZ57i",
        "outputId": "53805d29-c157-4ce0-a29d-ec5cbfdafff2"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "0\n",
            "1\n",
            "2\n",
            "3\n",
            "4\n",
            "5\n",
            "6\n",
            "7\n",
            "8\n",
            "9\n"
          ]
        }
      ],
      "source": [
        "for i in range(10):\n",
        "    print(i)\n",
        ""
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "base",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.12.1"
    },
    "colab": {
      "provenance": []
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}